x/incentive/keeper/msg_server_earn_test.go

```diff
@@ -0,0 +1,170 @@
+	valAddr1 := sdk.ValAddress(validatorAddr1)
+	valAddr2 := sdk.ValAddress(validatorAddr2)
+
+	bkavaDenom1 := fmt.Sprintf("bkava-%s", valAddr1.String())
+	bkavaDenom2 := fmt.Sprintf("bkava-%s", valAddr2.String())
+
+	authBuilder := suite.authBuilder().
+		WithSimpleAccount(userAddr1, cs(c("ukava", 1e12))).
+		WithSimpleAccount(userAddr2, cs(c("ukava", 1e12))).
+		WithSimpleAccount(validatorAddr1, cs(c("ukava", 1e12))).
+		WithSimpleAccount(validatorAddr2, cs(c("ukava", 1e12)))
+
+	incentBuilder := suite.incentiveBuilder()
+
+	savingsBuilder := testutil.NewSavingsGenesisBuilder().
+		WithSupportedDenoms("bkava")
+
+	earnBuilder := suite.earnBuilder().
+		WithVault(earntypes.AllowedVault{
+			Denom:             "bkava",
+			Strategies:        earntypes.StrategyTypes{earntypes.STRATEGY_TYPE_SAVINGS},
+			IsPrivateVault:    false,
+			AllowedDepositors: nil,
+		})
+
+	suite.SetupWithGenState(authBuilder, incentBuilder, earnBuilder, savingsBuilder)
+
+	err := suite.App.FundModuleAccount(suite.Ctx, distrtypes.ModuleName, cs(c("ukava", 1e12)))
+	suite.NoError(err)
+
+	// Create validators
+	err = suite.DeliverMsgCreateValidator(valAddr1, c("ukava", 1e9))
+	suite.Require().NoError(err)
+
+	err = suite.DeliverMsgCreateValidator(valAddr2, c("ukava", 1e9))
+	suite.Require().NoError(err)
+
+	// new block required to bond validator
+	suite.NextBlockAfter(7 * time.Second)
+	// Now the delegation is bonded, accumulate some delegator rewards
+	suite.NextBlockAfter(7 * time.Second)
+
+	// Create delegations from users
+	// User 1: 1e9 ukava to validator 1
+	// User 2: 99e9 ukava to validator 1 AND 2
+	err = suite.DeliverMsgDelegate(userAddr1, valAddr1, c("ukava", 1e9))
+	suite.Require().NoError(err)
+
+	err = suite.DeliverMsgDelegate(userAddr2, valAddr1, c("ukava", 99e9))
+	suite.Require().NoError(err)
+
+	err = suite.DeliverMsgDelegate(userAddr2, valAddr2, c("ukava", 99e9))
+	suite.Require().NoError(err)
+
+	// Mint liquid tokens
+	err = suite.DeliverMsgMintDerivative(userAddr1, valAddr1, c("ukava", 1e9))
+	suite.Require().NoError(err)
+
+	err = suite.DeliverMsgMintDerivative(userAddr2, valAddr1, c("ukava", 99e9))
+	suite.Require().NoError(err)
+
+	err = suite.DeliverMsgMintDerivative(userAddr2, valAddr2, c("ukava", 99e9))
+	suite.Require().NoError(err)
+
+	// Deposit liquid tokens to earn
+	err = suite.DeliverEarnMsgDeposit(userAddr1, c(bkavaDenom1, 1e9), earntypes.STRATEGY_TYPE_SAVINGS)
+	suite.Require().NoError(err)
+
+	err = suite.DeliverEarnMsgDeposit(userAddr2, c(bkavaDenom1, 99e9), earntypes.STRATEGY_TYPE_SAVINGS)
+	suite.Require().NoError(err)
+	err = suite.DeliverEarnMsgDeposit(userAddr2, c(bkavaDenom2, 99e9), earntypes.STRATEGY_TYPE_SAVINGS)
+	suite.Require().NoError(err)
+
+	// Accumulate some staking rewards
+	// _ = suite.App.EndBlocker(suite.Ctx, abci.RequestEndBlock{})
+	// suite.Ctx = suite.Ctx.WithBlockTime(time.Now().Add(1 * time.Hour))
+	// suite.App.BeginBlocker(suite.Ctx, abci.RequestBeginBlock{})
```

Comment:
nit: I'd maybe use the app begin end blockers so test matches production more closely, rather than calling the keeper methods to distribution etc. Was there problems with using endblockers?